### PR TITLE
[Feature] Add `capabilities` permissions to checkout extensions config

### PR DESCRIPTION
**Pull request by cpeddecord**
### WHY are these changes introduced?

Resolves internal Shopify issue, should be paired with: https://github.com/Shopify/shopify-cli-extensions/pull/288.

Thanks to @ecbrodie for pairing!

Adds support for a `capabilities` dictionary within an extension's `extension.config.yml`, this dict will expand but for now includes only the `network_access` with a boolean flag.

### WHAT is this pull request doing?

All changes in this PR are only relevant to Checkout Extensions and hence if the [Extensions Dev Server](https://github.com/Shopify/shopify-cli-extensions) is being used via enabling the `extension_server_beta` feature.

No other `shopify` commands should be affected by this change. There should also be no changes to the legacy checkout extensions endpoint, `/query`.

### How to test your changes?

<details>
<summary>Manual testing instructions</summary>

### Manual Testing Instructions

When running the below steps, make sure that you are using the local version of `shopify` rather than the production installation. The strategy I took was `alias shopifydev="path/to/shopify-cli/bin/shopify` and replace `shopify` with `shopifydev` for the commands below.

1. Follow environment setup instructions for both `shopify-cli` and `shopify-cli-extensions`. For Shopify employees, `dev up` will be useful for setting up both projects quickly.
1. Symlink the `shopify-cli-extensions` binary to `shopify-cli` via `rake extensions:symlink`
1. Enable the extensions dev server feature: `shopify config feature extension_server_beta --enable`
1. Create a new Checkout Extension via `shopify extension create`. Make sure you can select "Checkout Extension" as the type. This will require your Partner Organization and App having some beta flags enabled (refer to internal documentation or talk to @ecbrodie if you need help with this)
1. Modify `extension.config.yml` as necessary in your new extension (see below for test cases)
1. For each test case below, run `shopify extension serve` and observe the result in the extensions endpoint: `https://NGROK-TUNNEL-URL/extensions`. You will need to check the value of the `capabilities` list for the extension in the JSON. You also need to re-run this command for each test case because there is no hot reloading on config changes.

| Test Cases | Setup | Expected Results |
| ----------- | ------ | ----------------- |
| Test Case 1  | Set `capabilities: network_access:true` in `extension.config.yml`  |  Observe `"capabilities": {"network_access": true}` in the `/extensions` JSON |
| Test Case 2  | Omit `capabilities` in `extension.config.yml`  |  Observe `"capabilities": {"network_access": false}` in the `/extensions` JSON |
| Test Case 3 | Set `capabilities: { "network_access": "I shall throw"}` in `extension.config.yml`  |  `shopify extension serve` will error out (prepend the command with `SHOPIFY_CLI_STACKTRACE=1` to see the stack trace) |

</details>

### Post-release steps

* ship with https://github.com/Shopify/shopify-cli-extensions/pull/288.
* package the latest version of the `shopify-cli-extensions` binary in with `shopify-cli` during next release.

### Update checklist

- [x] I've added a CHANGELOG entry for this PR (if the change is public-facing)
- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows).
- [x] I've left the version number as is (we'll handle incrementing this when releasing).
- [x] I've included any post-release steps in the section above (if needed).

### CODE REVIEW

File: test/project_types/extension/tasks/convert_server_config_test.rb
Comment (ecbrodie) [inline test/project_types/extension/tasks/convert_server_config_test.rb:93]: As much as I like trailing commas, please revert this line, since we don't appear to have a linting rule in this project for trailing commas and this change is technically unrelated to the commit.

**Review by ecbrodie — COMMENTED** [review]
I paired with Chris, so we'll need one more code review before merge. I still need to tophat.

**Review by ecbrodie — APPROVED** [review]
Tophat was successful for me. Upon creating a new extension, I tested the following use cases of configuration in `extension.config.yml` and then running `shopify extension serve`:

- No `capabilities` (default upon creation): `/extensions` renders with no `"capabilities"` key in the extension
- Empty `capabilities` element in YML: `/extensions` renders with no `"capabilities"` key in the extension
- `capabilities` is an array of just `network_access`: `/extensions` renders with the `"capabilities"` array in the extension object included, and it contains the single string element of `"network_access"`
- `capabilities` is an array that has at least one value not equal to `'network_access'`: Appropriate validation error
- `capabilities` is a non-array value, such as `123`: Appropriate validation error

I had been hoping that the first two cases would still render `"capabilities": []` in the JSON rendered by `/extensions`, but they were omitted instead. This may be fine. I'll defer to @dmhenry on what is best in this case.

**Review by ecbrodie — APPROVED** [review]
I just did a retophat, with the rework of `capabilities` into an object structure. All test cases passed for me. I'm happy that there will always be a `capabilities` object and a `networkAccess` property present in the extensions JSON, even if either or both properties are omitted.

**Review by vinnylapenna — APPROVED** [review]
🎩 was successful, code LGTM

File: test/project_types/extension/tasks/convert_server_config_test.rb
Comment (cpeddecord) [reply test/project_types/extension/tasks/convert_server_config_test.rb:97]: rubocop prefers the `refute` here

**Pull request merged**
